dali/dfuplus/dfuplus.cpp

```diff
@@ -1448,7 +1452,262 @@ int CDfuPlusHelper::abort()
+
+    Owned<IClientListHistoryResponse> resp = dfuclient->ListHistory(req);
+
+    const IMultiException* excep = &resp->getExceptions();
+    if(excep != NULL && excep->ordinality() > 0)
+    {
+        StringBuffer errmsg;
+        excep->errorMessage(errmsg);
+        error("%s\n", errmsg.str());
+        return -1;
+    }
+
+    typedef enum
+    {
+        xml = 0,
+        ascii = 1,
+        csv = 2,
+        json = 3,
+    }out_t;
+
+    out_t format;
+
+    const char *formatPar = globals->queryProp("outformat");
+    if(formatPar == nullptr)
+        format = xml;
+    else if (stricmp(formatPar, "csv") == 0)
+        format=csv;
+    else if (stricmp(formatPar, "xml") == 0)
+        format=xml;
+    else if (stricmp(formatPar, "json") == 0)
+        format=json;
+    else if (stricmp(formatPar, "ascii") == 0)
+        format=ascii;
+    else
+    {
+        error("Unsupported output format: '%s'\n",formatPar);
+        return -1;
+    }
+
+    MemoryBuffer tmp;
+    tmp.append(resp->getXmlmap());
+    if (0 == *tmp.toByteArray())
+    {
+        error("%s doesn't have stored history!\n", lfn);
+        return -2;
+    }
+
+    Owned<IPropertyTree>    history;
+    history.setown(createPTree("History"));
+    history->deserialize(tmp);
+    if (!history->hasProp("Origin"))
+    {
+        error("%s doesn't have stored history!\n", lfn);
+        return -2;
+    }
+
+    bool sorted = true;
+
+    switch (format)
+    {
+    case xml:
+        {
+            StringBuffer out;
+            toXML(history, out, true);
+            info("\n%s\n",out.trim().str());
+        }
+        break;
+
+    case ascii:
+        {
+            progress("\nHistory of %s:\n", lfn);
+            unsigned index = 1;
+            Owned<IPropertyTreeIterator> historyIter = history->getElements("*");
+            ForEach(*historyIter)
+            {
+                info("%d", index++);
+                Owned<IAttributeIterator> attribIter = historyIter->query().getAttributes(sorted);
+                ForEach(*attribIter)
+                {
+                    info(", ");
+                    const char *propName = attribIter->queryName();
+                    if (*propName == '@')
+                        propName++;
+                    const char *propVal =  attribIter->queryValue();
+                    info("%s=%s",propName, propVal);
+                }
+                info("\n");
+            }
+        }
+        break;
+
+    case csv:
+        {
+            // TODO We need more sophisticated method if the record structure change in the future
+            bool showCsvHeader = globals->getPropBool("csvheader", true);
+            info("\n");
+            unsigned index = 1;
+            Owned<IPropertyTreeIterator> historyIter = history->getElements("*");
+            bool first = true;
+            if (showCsvHeader)
+            {
+                // Get header from the first record
+                (*historyIter).first();
```

Comment:
That's because it's a macro and bracketing to encapsulate anything that might be passed to it.